core/corehttp/gateway_handler.go

```diff
@@ -161,38 +175,97 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	// Resolve path to the final DAG node for the ETag
-	resolvedPath, err := i.api.ResolvePath(r.Context(), parsedPath)
-	switch err {
-	case nil:
-	case coreiface.ErrOffline:
-		webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusServiceUnavailable)
-		return
-	default:
-		webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
-		return
+	var (
+		resolvedPath              ipath.Resolved
+		rootPath                  string
+		escapedRootPath           string
+		resolvedRootPath          ipath.Resolved
+		dr                        files.Node
+		err                       error
+		isReedSolomonSubdirOrFile bool
+	)
+	top, err := i.isTopLevelEntryPath(r, parsedPath.String())
+	if err != nil {
+		switch err {
+		case nil:
+		case coreiface.ErrOffline:
+			webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusServiceUnavailable)
+			return
+		default:
+			webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
+			return
+		}
 	}
+	if !top {
+		// Check if the `parsedPath` is part of Reed-Solomon enconded directory object.
+		// If yes, put a new entry to the Reed-Solomon directory cache if not exist and
+		// get the files.Node of the `parsedPath`.
+		k := strings.SplitN(parsedPath.String(), "/", 4)[2]
+		v, isPresent, err := i.cacheEntryFor(k)
+		if err != nil {
+			internalWebError(w, err)
+			return
+		}
+		var rootDir files.Directory
+		escapedRootPath, err = getDirRootPath(escapedURLPath)
+		if err != nil {
+			webError(w, "invalid btfs path without slashes", err, http.StatusBadRequest)
+			return
+		}
+		if isPresent {
+			resolvedRootPath, rootDir = v.rootPath, v.rootDir
+		} else {
+			rootPath, err = getDirRootPath(parsedPath.String())
+			if err != nil {
+				webError(w, "invalid btfs path without slashes", err, http.StatusBadRequest)
+				return
+			}
 
-	dr, err := i.api.Unixfs().Get(r.Context(), resolvedPath)
-	if err != nil {
-		webError(w, "btfs cat "+escapedURLPath, err, http.StatusNotFound)
-		return
+			var rootDr files.Node
+			resolvedRootPath, rootDr, err = i.resolveAndGetFilesNode(context.Background(), w, r, ipath.New(rootPath), escapedRootPath)
+			if err != nil {
+				return
+			}
+			ok := false
+			rootDir, ok = rootDr.(files.Directory)
+			if !ok {
+				log.Warningf("expected directory type for %s", escapedRootPath)
+			}
+		}
+		if isPresent || rootDir != nil && rootDir.IsReedSolomon() {
+			if !isPresent {
+				// Put the current Reed-Solomon directory entry to the cache
+				i.rsDirs.Put(k, ReedSolomonDirectory{rootPath: resolvedRootPath, rootDir: rootDir})
+			}
+			// Get the files.Node of the `parsedPath`.
+			dr, err = findNode(rootDir, parsedPath.String())
+			if err != nil {
+				internalWebError(w, err)
+				return
+			}
+			isReedSolomonSubdirOrFile = true
+		}
+	}
+
+	if !isReedSolomonSubdirOrFile {
+		resolvedPath, dr, err = i.resolveAndGetFilesNode(r.Context(), w, r, parsedPath, escapedURLPath)
+		if err != nil {
+			return
```

Comment:
webError?

Reply:
The same as the above call. Please let me know if you have a different idea.